dataset-generation/spack_db/spack_db.py

```diff
@@ -0,0 +1,633 @@
+    if package_zip_hash is None:
+        print(f"No Tarball hash found in manifest: {package_filename}")
+
+        # Track taht this manifest has no downloadable binary tarball
+        with open(MISSING_TARBALL_HASH_FILE, "a") as f:
+            f.write(f"{package_hash}\n")
+        # go back to run_program function
+        return
+    
+    # track if the tarball hash has already been processed
+    if package_zip_hash in seen_tarball_hashes:
+        with open(SHARED_TARBALL_HASH_FILE, "a") as f:
+            # if this manifest points to a tarball that has already been seen,
+                # it will not create a new tarinfo entry
+                # it will have a new manifest entry
+            f.write(f"{package_hash}\t{package_zip_hash}\n")
+    else:
+        seen_tarball_hashes.add(package_zip_hash)
+        
+
+    expected_tarinfo_hash = package_zip_hash
+
+    if expected_tarinfo_hash in existing_tarinfo_files:
+        print(f"✅ Already have tarinfo for {name}-{version}-{expected_tarinfo_hash}, skipping binary download.")
+        update_index_entry(index, package_hash, package_value, package_zip_hash)
+        save_index(index)
+        print(f"✅ Saved Index")
+        return
+    
+    else:
+
+        # make the binary package URL for installing the package
+        binary_package_URL = make_binary_package_URL(package_zip_hash)
+        print(f"🔗 Downloading binary: {binary_package_URL}")
+
+        # download the binary package file from the generated URL
+        tempbinary = download_from_URL(binary_package_URL, package_filename, is_spec = False)
+        print("✅ Binary package downloaded")
+
+        if tempbinary is None:
+
+            # Track failed tarball download
+            with open(FAILED_TARBALL_DOWNLOAD_FILE, "a") as f:
+                f.write(f"{package_filename}: manifest hash: {package_hash}, tarball hash: {package_zip_hash}\n")
+            
+            return
+
+        # read the binary package
+        read_binary_package(tempbinary, package_filename, package_zip_hash)
+        print("✅ Finished reading binary package")
+        update_index_entry(index, package_hash, package_value, package_zip_hash)
+        save_index(index)
+        print(f"Updated Index with {package_filename}-{package_zip_hash}")
+        save_checkpoint(package_hash)
+        print(f"✅ Saved Index")
+
+# program dispatcher
+def run_program(package_hash, database, index, existing_tarinfo_files, seen_tarball_hashes): 
+    installs = database['database']['installs']
+
+    # gets installs key value, aka name of package, version, etc.
+    package_value = installs[package_hash]
+    print_files(package_hash, package_value, index, existing_tarinfo_files, seen_tarball_hashes)
+
+
+def main():
+    #file_name = "myMedjson.json"
+    # file_name = "myjson.json"
+    # file_name = 'Med_w_compilerwrapper_packages_at_end.json'
+    file_name = "e2a6969c742c8ee33deba2d210ce2243cd3941c6553a3ffc53780ac6463537a9"
```

Comment:
Making your script take the file name as an argument would make it more flexible. Or making it also fetch the latest version of the spack index db file (e.g. that happened to have the name e2a6969c7 when we originally looked at how the spack package repository stores stuff).